Casing and expand (#417)
Microsoft/api-guidelines: graph/patterns/change-tracking.md

# Change tracking

Microsoft Graph API Design Pattern

*The change tracking pattern provides the ability for API consumers to request changes in data from Microsoft Graph without having to re-read data that has not changed.*

## Problem

API consumers require an efficient way to acquire changes to data in the Microsoft Graph, for example to synchronize an external store or to drive a change-centric business process.

## Solution

> API designers can enable the change tracking (delta) capability on a resource in the Microsoft Graph (typically on an entity collection or a parent resource) by declaring a delta function on that resource.

Author: @garethj-msft   IMO we need to clarify that change tracking is enabled by declaring a delta function and applying ChangeTracking annotation to the targeted resource.

Reviewer: We're also about to release support for change tracking on a function - but I guess we can update the doc once that's officially released.

Author: Add ChangeTracking annotation

This function returns a delta payload. A delta payload consists of a collection of annotated full or partial Microsoft Graph entities plus either a `nextLink` to further pages of original or change data that are immediately available OR a `deltaLink` to poll to get the next set of changes as they occur.

Annotations allow the delta payload to indicate resources or links which have been deleted. API callers are expected to differentiate resource adds from updates by interpreting the id property of the change records against the existence of resources in whatever external system is doing the processing.

> The pattern requires a sequence of requests.

Reviewer: The pattern bellow doesn't seem correct.  Please see the details here: https://learn.microsoft.com/en-us/graph/delta-query-overview?tabs=http#use-delta-query-to-track-changes-in-a-resource-collection

Author: Maybe I'm missing something, but I believe it's very close: 1. delta no param, 2. optional if there is nextLink 3. update on delta with delta token 4. optional if there is nextLink. added a link to pub docs

Author: @dkershaw10 could you please elaborate on what is the main deviation besides ability to use latest?

  - There is no mechanism to indicate that a resource has entered or exited the dataset based on a change that causes it to match or no longer match any `$filter` query parameter.
  - When a link to an entity is deleted, when the linked entity is deleted, or when a link to an entity is added, the implementer MUST return a `property@delta` annotation. 
    - When a link to an entity is deleted, but the entity still exists, the reason MUST be set to `changed`.
    - When a link to an entity is deleted along with the entity, the reason MUST be set to `deleted`.
  
API producers MAY choose to collate multiple changes to the same resource into a single change record. 

## When to use this pattern

API consumers want a pull mechanism to request and process change to Microsoft Graph data, either via proactive polling or by responding to Microsoft Graph notifications.

API consumers need guaranteed data integrity over the set of changes to Microsoft Graph data.

## Issues and considerations

 - API service MAY be able to respond to standard OData query parameters with the initial call to the delta function:

    - `$select` to enforce the set of properties on which change is reported.

>     - `$filter` to influence the range of changes returned.

Reviewer: "range" or "scope"?  Range makes it sound like a duration range, but we have other supported filters that are not duration based.

- API producers might use `$skipToken` and `$deltaToken` within their implementations of `nextLink` and `deltaLink`, however the URLs are defined as being opaque and the existence of the tokens MUST NOT be documented.   It is not a breaking change to modify the structure of `nextLinks` or `deltaLinks`.- 
- `nextLink` and `deltaLink` URLs are valid for a specific period before the client application needs to run a full synchronization again.For `nextLink`, a minimal validity time should be 1 hour. For `deltaLink`, a minimal validity time should be seven days. When a link is no longer valid it must return a standard error with a 410 GONE response code.
  

## Alternatives

- Change notifications pattern with rich payloads – for use cases where API consumers would find calling back into Microsoft Graph onerous and absolute integrity guarantees are less critical.

## Examples

> ### Change tracking on entity set

Author: @garethj-msft I think we need to add an example of ChangeTracking on an Entities as it is one of most used patterns

    <Annotation Term="Org.OData.Capabilities.V1.ChangeTracking"> 
      <Record> 
        <PropertyValue Property="Supported" Bool="true" /> 
      </Record> 
    </Annotation> 
</EntitySet> 
```

### Change tracking on navigation property

```
<Function Name="delta" IsBound="true">
        <Parameter Name="bindingParameter" Type="Collection(microsoft.education.rostering.api.educationClass)" />
        <ReturnType Type="Collection(microsoft.education.rostering.api.educationClass)" />
      </Function>
<EntityType Name="educationRoot">
        <NavigationProperty Name="classes" Type="Collection(microsoft.education.rostering.api.educationClass)" ContainsTarget="true">
          <Annotation Term="Org.OData.Capabilities.V1.ChangeTracking">
            <Record>
              <PropertyValue Property="Supported" Bool="true" />
            </Record>
          </Annotation>
        </NavigationProperty>
</EntityType>
```

### Delta payload

 Here, a user resource is updated, and there is one user added to and one removed from that user’s directReports collection. Additionally, a second user is deleted. In this case, there are no further pages of change records currently available.

```

> GET https://graph.microsoft.com/v1.0/users/delta

Reviewer: Shouldn't this request have a $deltaToken (based on the $odata.deltaLink property returned previously) as you are showing deltas in the response?

# Change tracking

Microsoft Graph API Design Pattern

*The change tracking pattern provides the ability for API consumers to request changes in data from Microsoft Graph without having to re-read data that has not changed.*

## Problem

API consumers require an efficient way to acquire changes to data in the Microsoft Graph, for example to synchronize an external store or to drive a change-centric business process.

## Solution

API designers can enable the change tracking (delta) capability on a resource in the Microsoft Graph (typically on an entity collection or a parent resource) by declaring a delta function on that resource and applying `Org.OData.Capabilities.V1.ChangeTracking` annotation.

> This function returns a delta payload. A delta payload consists of a collection of annotated full or partial Microsoft Graph entities plus either a `nextLink` to further pages of original or change data that are immediately available OR a `deltaLink` to poll to get the next set of changes as they occur.

Reviewer: ```suggestion
This function returns a delta payload. A delta payload consists of a collection of annotated full or partial Microsoft Graph entities plus either a `nextLink` to further pages of original or change data that are immediately available OR a `deltaLink` to get the next set of changes at some later date.
```

Annotations allow the delta payload to indicate resources or links which have been deleted. API callers are expected to differentiate resource adds from updates by interpreting the id property of the change records against the existence of resources in whatever external system is doing the processing.

The pattern requires a sequence of requests on the delta function, for additional details see [Change Tracking](https://learn.microsoft.com/en-us/graph/delta-query-overview?tabs=http#use-delta-query-to-track-changes-in-a-resource-collection):

  1. GET request which returns the first page of the current state of the resources that delta applies to.  
  2. [Optionally] Further GET requests to retrieve more pages of the current state via the `@odata.nextLink` URL.
  3. After some time, a GET request to see if there are new changes via the `@odata.deltaLink` URL.
  4. [Optionally] GET requests to retrieve more pages of changes via the `@odata.nextLink` URL.

> The `nextLink` provides a mechanism to do server-driven paging through the change data that is currently available.  When there are no further pages of changes immediately available, a `deltaLink` is returned instead.

Reviewer: I think you should move lines 27 and 29 up before line 18.  That provide better context about nextLink and deltaLink that makes the sequence described in the enumerated list above clearer.

> The `deltaLink` provides a mechanism for the API consumer to catch up on changes since their last request to the delta function, `deltaLink` or last page’s `nextLink`. If no changes have happened since the last request, then the deltaLink MUST return an empty collection.

Reviewer: ```suggestion
The `deltaLink` provides a mechanism for the API consumer to catch up on changes since their last request to the delta function. If no changes have happened since the last request, then the deltaLink MUST return an empty collection.
```

    - `$select` to enforce the set of properties on which change is reported.
    - `$filter` to influence the scope of changes returned.
    - `$expand` to include linked resources with the set of changes.
    - `$top` parameter to influence the size of the set of change records.
  
    These query parameters MUST be encoded into subsequent `@odata.nextLink` or `@odata.deltaLink`, such that the same options are preserved through the call sequence without callers respecifying them, which MUST NOT be allowed. OData query parameters must be honored in full, or a 400-error returned.
- Making a sequence of calls to a delta function followed by the opaque URLs in the `nextLink` and `deltaLink` MUST guarantee that the data at the start time of the call sequence and all changes to the data thereafter will be returned at least once. It is not necessary to avoid duplicates in the sequence. When the delta function is returning changes, they MUST be sequenced chronologically refer to [public documentation](https://learn.microsoft.com/en-us/graph/delta-query-overview?view=graph-rest-1.0) for more details.

> - The delta function can be bound to a collection, as with

Reviewer: I can also be bound to a function that returns an entity collection.  See [Pull Request 7242963](https://dev.azure.com/msazure/One/_git/AD-AggregatorService-Wiki/pullrequest/7242963): Add info on change tracking for functions, for an example, including the schema updates.

Author: Actually the "some logical parent resource" phrase allows for a function but i also added an example and formatting to highlight options.

`/users/delta` that returns the changes to the users' collection
  or to some logical parent resource, where the change records are implied to be relative to all collections contained within the parent, for example
  `/me/planner/all/delta` – this returns changes to any resource within planner that a user is subscribed to as a heterogenous collection.

> - API producers might use `$skipToken` and `$deltaToken` within their implementations of `nextLink` and `deltaLink`, however the URLs are defined as being opaque and the existence of the tokens MUST NOT be documented.   It is not a breaking change to modify the structure of `nextLinks` or `deltaLinks`.- 

Reviewer: I'm not sure if this is true.  AGS looks for and encodes/decodes the skip and delta tokens. I think this is a *should/must* rather than *might*.

- `nextLink` and `deltaLink` URLs are valid for a specific period before the client application needs to run a full synchronization again.For `nextLink`, a minimal validity time should be 1 hour. For `deltaLink`, a minimal validity time should be seven days. When a link is no longer valid it must return a standard error with a 410 GONE response code.
- Although this capability is similar to the OData `$delta` feed capability, it is a different construct. Microsoft Graph APIs MUST provide change tracking through the delta function and MUST NOT implement the OData `$delta` feed when providing change tracking capabilities to ensure the uniformity of the API experience.
- The Graph delta payload format has some deviations from the OData 4.01 change tracking format to simplify parsing, for example the context annotation is removed.
  

## Alternatives

- Change notifications pattern with rich payloads – for use cases where API consumers would find calling back into Microsoft Graph onerous and absolute integrity guarantees are less critical.

## Examples

### Change tracking on entity set

```

> <Function Name="delta" IsBound="true" ags:OwnerService="Microsoft.DirectoryServices">

Reviewer: Should these be the **public** CSDLs rather than workload CSDLs?
If so remove any ags annotations, and coerce to the microsoft.graph namespace for type names.

Author: updated
